### ISSUE TITLE
Corrupted Active Storage attachments

### ISSUE THREAD

**Issue by kennyadsl** [solidusio/solidus#4017]
Hey guys!
I have an issue using `ActiveStorageAttachment` for product images.
I'm using an S3-compatible service for storing attachments, and it's working ok. But if the service returns an error (it is down at the moment the user attempts to upload an image or there's a permission error, etc), a product image record and ActiveStorage blob record are created anyway despite the image processing error. So it produces a corrupted state of the system. And then any code which calls `.url(style)` on this image gets an `ActiveStorage::FileNotFoundError` exception, because it tries to run image processing to create the requested image variant, but it can't find the original image because its upload never finished (it happens on any page which needs to display the image).
Ideally, I would like it to be "all or nothing", so the database records get committed only after the storage service responded with a success.
Is it a problem of how ActiveStorage works itself or is it a Solidus wrapper's one? How do I prevent the situation when one failed image upload breaks down the whole store (you can't even delete the corrupted image, because the admin product page is also broken)?

_Originally posted by @ok32 in https://github.com/solidusio/solidus/discussions/3986_

**Comment by kennyadsl** [comment]
I had another issue uploading a not supported file type (`.heic`), not sure it's related but I also had to remove the image from the database to fix it. To reproduce:

```
rails new store --skip-javascript
cd store
bundle add solidus -v 3.0.0.rb2
bin/rails generate solidus:install
```

Login into the admin and upload a `.heic` file in any product. Reload the page to see the `ActiveStorage::InvariableError`.

**Comment by cpfergus1** [comment]
I was able recreate this issue by:
1. creating a new application
2.  add `gem 'solidus', path: '/Users/*_user_*/solidus-3` and:
```
$ bundle install
$ bin/rails generate solidus:install
```
3. setup s3 bucket at AWS (https://docs.aws.amazon.com/AmazonS3/latest/userguide/creating-bucket.html)
4. change root/config/storage.yml to include s3 bucket:
```amazon:
  service: S3
  access_key_id: <%= Rails.application.credentials.dig(:aws, :access_key_id) %>
  secret_access_key: <%= Rails.application.credentials.dig(:aws, :secret_access_key) %>
  region: us-west-2
  bucket: *_bucket_*```
5. change root/config/environments/development.rb `config.active_storage.service = :amazon`
6. slightly change your AWS credentials prior to booting up your server by commenting out part of your ID
7. Boot up server, and navigate to admin page to try and upload a photo - should receive uploading error.
8. reload webpage and get first error Aws::S3::Errors::InvalidAccessKeyId
9. reload server with correct credentials and reload page to receive the desired error ActiveStorage::FileNotFoundError in Spree::Admin::Images#index

The issue appears to be with ActiveStorage as it queues the upload of the image after the objects are created and when it fails to upload, it does not purge the existence of the image and the created objects, image and attachment, persist with an ActiveStorage record still found in the database. Subsequent calls for the object expect a file to be returned from the server and solidus currently does not handle the exception for missing file. 

Initial remedy provided by the original poster was an exception handle in ActiveStorageAttachment that appended a revised url method:

```
    def url(style = nil)
      super(style)
    rescue ActiveStorage::FileNotFoundError
      '/images/corrupted-image.png'
    end
```
This solution does not rectify the creation of an image or attachment when there is a failure to upload, and only bypasses an exception if the the image cannot be found on third part storage.

Ideally we would want to prevent the object from ever being created should the image fail to upload. This might be difficult to accomplish based on the way the the objects and jobs are created. Therefore, we could also potentially run a validation ActiveStorageAdapter or implement a post processing job to clean up Images with failed uploads.


Possible remedies may include

**Comment by cpfergus1** [comment]
Adding stack trace for further investigation
`activestorage (6.1.3.1) lib/active_storage/service/s3_service.rb:139:in `stream'
activestorage (6.1.3.1) lib/active_storage/service/s3_service.rb:41:in `block in download'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
activestorage (6.1.3.1) lib/active_storage/service.rb:155:in `instrument'
activestorage (6.1.3.1) lib/active_storage/service/s3_service.rb:40:in `download'
activestorage (6.1.3.1) lib/active_storage/downloader.rb:32:in `download'
activestorage (6.1.3.1) lib/active_storage/downloader.rb:13:in `block in open'
activestorage (6.1.3.1) lib/active_storage/downloader.rb:24:in `open_tempfile'
activestorage (6.1.3.1) lib/active_storage/downloader.rb:12:in `open'
activestorage (6.1.3.1) lib/active_storage/service.rb:90:in `open'
activestorage (6.1.3.1) app/models/active_storage/blob.rb:276:in `open'
activestorage (6.1.3.1) app/models/active_storage/variant_with_record.rb:34:in `transform_blob'
activestorage (6.1.3.1) app/models/active_storage/variant_with_record.rb:16:in `process'
activestorage (6.1.3.1) app/models/active_storage/variant_with_record.rb:11:in `processed'
/Users/cpfergus1/nebulab_solidus/solidus/core/app/models/concerns/spree/active_storage_adapter/attachment.rb:34:in `variant'
/Users/cpfergus1/nebulab_solidus/solidus/core/app/models/concerns/spree/active_storage_adapter/attachment.rb:26:in `url'
/Users/cpfergus1/nebulab_solidus/solidus/core/app/models/concerns/spree/active_storage_adapter.rb:110:in `url'
activesupport (6.1.3.1) lib/active_support/core_ext/object/try.rb:15:in `public_send'
activesupport (6.1.3.1) lib/active_support/core_ext/object/try.rb:15:in `try'
/Users/cpfergus1/nebulab_solidus/solidus/backend/app/views/spree/admin/shared/_image.html.erb:3
actionview (6.1.3.1) lib/action_view/base.rb:247:in `public_send'
actionview (6.1.3.1) lib/action_view/base.rb:247:in `_run'
actionview (6.1.3.1) lib/action_view/template.rb:154:in `block in render'
activesupport (6.1.3.1) lib/active_support/notifications.rb:205:in `instrument'
actionview (6.1.3.1) lib/action_view/template.rb:345:in `instrument_render_template'
actionview (6.1.3.1) lib/action_view/template.rb:152:in `render'
actionview (6.1.3.1) lib/action_view/renderer/partial_renderer.rb:285:in `block in render_partial_template'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
actionview (6.1.3.1) lib/action_view/renderer/partial_renderer.rb:280:in `render_partial_template'
actionview (6.1.3.1) lib/action_view/renderer/partial_renderer.rb:271:in `render'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:81:in `render_partial_to_object'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:53:in `render_partial'
actionview (6.1.3.1) lib/action_view/helpers/rendering_helper.rb:45:in `render'
/Users/cpfergus1/nebulab_solidus/solidus/backend/app/views/spree/admin/images/_image_row.html.erb:11
actionview (6.1.3.1) lib/action_view/base.rb:247:in `public_send'
actionview (6.1.3.1) lib/action_view/base.rb:247:in `_run'
actionview (6.1.3.1) lib/action_view/template.rb:154:in `block in render'
activesupport (6.1.3.1) lib/active_support/notifications.rb:205:in `instrument'
actionview (6.1.3.1) lib/action_view/template.rb:345:in `instrument_render_template'
actionview (6.1.3.1) lib/action_view/template.rb:152:in `render'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:189:in `block in collection_with_template'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:70:in `block in each_with_info'
activerecord (6.1.3.1) lib/active_record/relation/delegation.rb:88:in `each'
activerecord (6.1.3.1) lib/active_record/relation/delegation.rb:88:in `each'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:70:in `each_with_info'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:88:in `each_with_info'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:180:in `each'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:180:in `map'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:180:in `collection_with_template'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:162:in `block (2 levels) in render_collection'
actionview (6.1.3.1) lib/action_view/renderer/partial_renderer/collection_caching.rb:21:in `cache_collection_render'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:161:in `block in render_collection'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:147:in `render_collection'
actionview (6.1.3.1) lib/action_view/renderer/collection_renderer.rb:119:in `render_collection_with_partial'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:72:in `render_partial_to_object'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:27:in `render_to_object'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:22:in `render'
actionview (6.1.3.1) lib/action_view/helpers/rendering_helper.rb:38:in `block in render'
actionview (6.1.3.1) lib/action_view/base.rb:273:in `in_rendering_context'
actionview (6.1.3.1) lib/action_view/helpers/rendering_helper.rb:34:in `render'
/Users/cpfergus1/nebulab_solidus/solidus/backend/app/views/spree/admin/images/index.html.erb:70
actionview (6.1.3.1) lib/action_view/base.rb:247:in `public_send'
actionview (6.1.3.1) lib/action_view/base.rb:247:in `_run'
actionview (6.1.3.1) lib/action_view/template.rb:154:in `block in render'
activesupport (6.1.3.1) lib/active_support/notifications.rb:205:in `instrument'
actionview (6.1.3.1) lib/action_view/template.rb:345:in `instrument_render_template'
actionview (6.1.3.1) lib/action_view/template.rb:152:in `render'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:61:in `block (2 levels) in render_template'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:56:in `block in render_template'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:71:in `block in render_with_layout'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:70:in `render_with_layout'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:55:in `render_template'
actionview (6.1.3.1) lib/action_view/renderer/template_renderer.rb:11:in `render'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:61:in `render_template_to_object'
actionview (6.1.3.1) lib/action_view/renderer/renderer.rb:29:in `render_to_object'
actionview (6.1.3.1) lib/action_view/rendering.rb:117:in `block in _render_template'
actionview (6.1.3.1) lib/action_view/base.rb:273:in `in_rendering_context'
actionview (6.1.3.1) lib/action_view/rendering.rb:116:in `_render_template'
actionpack (6.1.3.1) lib/action_controller/metal/streaming.rb:218:in `_render_template'
actionview (6.1.3.1) lib/action_view/rendering.rb:103:in `render_to_body'
actionpack (6.1.3.1) lib/action_controller/metal/rendering.rb:52:in `render_to_body'
actionpack (6.1.3.1) lib/action_controller/metal/renderers.rb:142:in `render_to_body'
actionpack (6.1.3.1) lib/abstract_controller/rendering.rb:25:in `render'
actionpack (6.1.3.1) lib/action_controller/metal/rendering.rb:36:in `render'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:46:in `block (2 levels) in render'
/Users/cpfergus1/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/benchmark.rb:308:in `realtime'
activesupport (6.1.3.1) lib/active_support/core_ext/benchmark.rb:14:in `ms'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:46:in `block in render'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:86:in `cleanup_view_runtime'
activerecord (6.1.3.1) lib/active_record/railties/controller_runtime.rb:34:in `cleanup_view_runtime'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:45:in `render'
actionpack (6.1.3.1) lib/action_controller/metal/implicit_render.rb:35:in `default_render'
actionpack (6.1.3.1) lib/action_controller/metal/basic_implicit_render.rb:6:in `send_action'
actionpack (6.1.3.1) lib/abstract_controller/base.rb:228:in `process_action'
actionpack (6.1.3.1) lib/action_controller/metal/rendering.rb:30:in `process_action'
actionpack (6.1.3.1) lib/abstract_controller/callbacks.rb:42:in `block in process_action'
activesupport (6.1.3.1) lib/active_support/callbacks.rb:117:in `block in run_callbacks'
actiontext (6.1.3.1) lib/action_text/rendering.rb:20:in `with_renderer'
actiontext (6.1.3.1) lib/action_text/engine.rb:55:in `block (4 levels) in <class:Engine>'
activesupport (6.1.3.1) lib/active_support/callbacks.rb:126:in `instance_exec'
activesupport (6.1.3.1) lib/active_support/callbacks.rb:126:in `block in run_callbacks'
activesupport (6.1.3.1) lib/active_support/callbacks.rb:137:in `run_callbacks'
actionpack (6.1.3.1) lib/abstract_controller/callbacks.rb:41:in `process_action'
actionpack (6.1.3.1) lib/action_controller/metal/rescue.rb:22:in `process_action'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:34:in `block in process_action'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `block in instrument'
activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
activesupport (6.1.3.1) lib/active_support/notifications.rb:203:in `instrument'
actionpack (6.1.3.1) lib/action_controller/metal/instrumentation.rb:33:in `process_action'
actionpack (6.1.3.1) lib/action_controller/metal/params_wrapper.rb:249:in `process_action'
activerecord (6.1.3.1) lib/active_record/railties/controller_runtime.rb:27:in `process_action'
actionpack (6.1.3.1) lib/abstract_controller/base.rb:165:in `process'
actionview (6.1.3.1) lib/action_view/rendering.rb:39:in `process'
actionpack (6.1.3.1) lib/action_controller/metal.rb:190:in `dispatch'
actionpack (6.1.3.1) lib/action_controller/metal.rb:254:in `dispatch'
actionpack (6.1.3.1) lib/action_dispatch/routing/route_set.rb:50:in `dispatch'
actionpack (6.1.3.1) lib/action_dispatch/routing/route_set.rb:33:in `serve'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:50:in `block in serve'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `each'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `serve'
actionpack (6.1.3.1) lib/action_dispatch/routing/route_set.rb:842:in `call'
railties (6.1.3.1) lib/rails/engine.rb:539:in `call'
railties (6.1.3.1) lib/rails/railtie.rb:207:in `public_send'
railties (6.1.3.1) lib/rails/railtie.rb:207:in `method_missing'
actionpack (6.1.3.1) lib/action_dispatch/routing/mapper.rb:20:in `block in <class:Constraints>'
actionpack (6.1.3.1) lib/action_dispatch/routing/mapper.rb:49:in `serve'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:50:in `block in serve'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `each'
actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `serve'
actionpack (6.1.3.1) lib/action_dispatch/routing/route_set.rb:842:in `call'
warden (1.2.9) lib/warden/manager.rb:36:in `block in call'
warden (1.2.9) lib/warden/manager.rb:34:in `catch'
warden (1.2.9) lib/warden/manager.rb:34:in `call'
rack (2.2.3) lib/rack/tempfile_reaper.rb:15:in `call'
rack (2.2.3) lib/rack/etag.rb:27:in `call'
rack (2.2.3) lib/rack/conditional_get.rb:27:in `call'
rack (2.2.3) lib/rack/head.rb:12:in `call'
actionpack (6.1.3.1) lib/action_dispatch/http/permissions_policy.rb:22:in `call'
actionpack (6.1.3.1) lib/action_dispatch/http/content_security_policy.rb:18:in `call'
rack (2.2.3) lib/rack/session/abstract/id.rb:266:in `context'
rack (2.2.3) lib/rack/session/abstract/id.rb:260:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/cookies.rb:689:in `call'
activerecord (6.1.3.1) lib/active_record/migration.rb:601:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
activesupport (6.1.3.1) lib/active_support/callbacks.rb:98:in `run_callbacks'
actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
web-console (4.1.0) lib/web_console/middleware.rb:132:in `call_app'
web-console (4.1.0) lib/web_console/middleware.rb:28:in `block in call'
web-console (4.1.0) lib/web_console/middleware.rb:17:in `catch'
web-console (4.1.0) lib/web_console/middleware.rb:17:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
railties (6.1.3.1) lib/rails/rack/logger.rb:37:in `call_app'
railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `block in call'
activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `block in tagged'
activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:37:in `tagged'
activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `tagged'
railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `call'
sprockets-rails (3.2.2) lib/sprockets/rails/quiet_assets.rb:13:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/request_id.rb:26:in `call'
rack (2.2.3) lib/rack/method_override.rb:24:in `call'
rack (2.2.3) lib/rack/runtime.rb:22:in `call'
activesupport (6.1.3.1) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/static.rb:24:in `call'
rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
actionpack (6.1.3.1) lib/action_dispatch/middleware/host_authorization.rb:98:in `call'
rack-mini-profiler (2.3.1) lib/mini_profiler/profiler.rb:373:in `call'
railties (6.1.3.1) lib/rails/engine.rb:539:in `call'
puma (5.2.2) lib/puma/configuration.rb:248:in `call'
puma (5.2.2) lib/puma/request.rb:76:in `block in handle_request'
puma (5.2.2) lib/puma/thread_pool.rb:337:in `with_force_shutdown'
puma (5.2.2) lib/puma/request.rb:75:in `handle_request'
puma (5.2.2) lib/puma/server.rb:431:in `process_client'
puma (5.2.2) lib/puma/thread_pool.rb:145:in `block in spawn_thread'`

**Comment by cpfergus1** [comment]
Stack trace for further investigation

[StackTrace.txt](https://github.com/solidusio/solidus/files/6281890/StackTrace.txt)
